core/views.py

```diff
@@ -148,50 +183,77 @@ def setting(request):
     if request.method == "POST":
         hackathon = get_object_or_404(Hackathon, id=hackathon_id)
         hacker = get_object_or_404(Hacker, id=request.user.id)
+        page_type = request.GET.get("page_type")
 
         if hacker.saved.filter(id=hackathon_id).exists():
+            hacker.saved.remove(hackathon)
+            if page_type == "saved":
+                return redirect("saved_hackathons")
+            else:
+                return JsonResponse({"status": "sucess", "hackathon": "removed"})
+
+        hacker.saved.add(hackathon)
+        if page_type == "saved":
+            return redirect("saved_hackathons")
+        else:
             return JsonResponse(
-                {"status": "fail", "error": "Invalid request"}, status=400
+                {
+                    "status": "success",
+                    "hackathon": {
+                        "id": hackathon.id,
+                        "name": hackathon.name,
+                    },
+                }
             )
 
-        hacker.saved.add(hackathon)
-        return JsonResponse(
-            {
-                "status": "success",
-                "hackathon": {
-                    "id": hackathon.id,
-                    "name": hackathon.name,
-                },
-            }
-        )
 
+def str_to_bool(s):
+    return s.lower() in ["true", "1", "yes"]
 
-@login_required
-def unsave_hackathon(request: HttpRequest, hackathon_id):
+
+def add_vote(request: HttpRequest, hackathon_id):
     if request.method == "POST":
         hackathon = get_object_or_404(Hackathon, id=hackathon_id)
         hacker = get_object_or_404(Hacker, id=request.user.id)
+        type_vote = request.GET.get("vote_state")
+        state = str_to_bool(type_vote)
+        vote = Vote.objects.filter(hackathon_id=hackathon, from_hacker=hacker)
+
+        if vote.exists():
+            vote = vote.first()
+            if vote.type_vote:
+                hackathon.count_upvotes -= 1
+                hackathon.count_downvotes += 1
+            else:
+                hackathon.count_downvotes -= 1
+                hackathon.count_upvotes += 1
 
-        if not hacker.saved.filter(id=hackathon_id).exists():
-            return JsonResponse(
-                {"status": "fail", "error": "Invalid request"}, status=400
-            )
+            # Update the vote type
+            vote.type_vote = state
+            vote.save()
+        else:
+            vote = Vote.objects.create(hackathon_id=hackathon, type_vote=state)
 
-        hacker.saved.remove(hackathon)
-        return JsonResponse(
-            {
-                "status": "success",
-                "hackathon": {
-                    "id": hackathon.id,
-                    "name": hackathon.name,
-                },
-            }
-        )
+            vote.from_hacker.add(hacker)
+
+        hackathon.count_upvotes = Vote.objects.filter(
+            hackathon_id=hackathon, type_vote=True
```

Comment:
You'll want a votes field on the Hackathon model as this method is super slow  and non scaleable